File: Tellma.Model.Application/LineDefinitionEntry.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Tellma.Model.Common;

namespace Tellma.Model.Application
{
    [Display(Name = "LineDefinitionEntry", GroupName = "LineDefinitionEntries")]
    public class LineDefinitionEntryForSave<TAgentDef, TResourceDef, TNotedAgentDef> : EntityWithKey<int>
    {
        [Display(Name = "LineDefinitionEntry_Direction")]
        [ChoiceList(new object[] { 
                (short)1, 
                (short)-1 }, 
            new string[] { 
                "Entry_Direction_Debit", 
                "Entry_Direction_Credit" })]
        [Required, ValidateRequired]
        public short? Direction { get; set; }

        [Display(Name = "LineDefinitionEntry_ParentAccountType")]
        [Required, ValidateRequired]
        public int? ParentAccountTypeId { get; set; }

        [Display(Name = "LineDefinitionEntry_EntryType")]
        public int? EntryTypeId { get; set; }

        [Display(Name = "LineDefinitionEntry_AgentDefinitions")]
        [ForeignKey(nameof(LineDefinitionEntryAgentDefinition.LineDefinitionEntryId))]
        public List<TAgentDef> AgentDefinitions { get; set; }

        [Display(Name = "LineDefinitionEntry_ResourceDefinitions")]
        [ForeignKey(nameof(LineDefinitionEntryResourceDefinition.LineDefinitionEntryId))]
        public List<TResourceDef> ResourceDefinitions { get; set; }

        [Display(Name = "LineDefinitionEntry_NotedAgentDefinitions")]
        [ForeignKey(nameof(LineDefinitionEntryNotedAgentDefinition.LineDefinitionEntryId))]
        public List<TNotedAgentDef> NotedAgentDefinitions { get; set; }
    }

    public class LineDefinitionEntryForSave : LineDefinitionEntryForSave<LineDefinitionEntryAgentDefinitionForSave, LineDefinitionEntryResourceDefinitionForSave, LineDefinitionEntryNotedAgentDefinitionForSave>
    {
    }

    public class LineDefinitionEntry : LineDefinitionEntryForSave<LineDefinitionEntryAgentDefinition, LineDefinitionEntryResourceDefinition, LineDefinitionEntryNotedAgentDefinition>
    {
        [Required]
        public int? Index { get; set; }

        [Display(Name = "LineDefinitionEntry_LineDefinition")]
        [Required]
        public int? LineDefinitionId { get; set; }

        [Display(Name = "ModifiedBy")]
        [Required]
        public int? SavedById { get; set; }

        // For Query

        [Display(Name = "LineDefinitionEntry_ParentAccountType")]
        [ForeignKey(nameof(ParentAccountTypeId))]
        public AccountType ParentAccountType { get; set; }

        [Display(Name = "LineDefinitionEntry_EntryType")]
        [ForeignKey(nameof(EntryTypeId))]
        public EntryType EntryType { get; set; }

        [Display(Name = "ModifiedBy")]
        [ForeignKey(nameof(SavedById))]
        public User SavedBy { get; set; }
    }
}

File: Tellma.Api.Web/ClientApp/src/app/data/entities/line-definition-entry.ts
import { EntityForSave } from './base/entity-for-save';
import {
    LineDefinitionEntryAgentDefinition,
    LineDefinitionEntryAgentDefinitionForSave
} from './line-definition-entry-agent-definition';
import {
    LineDefinitionEntryResourceDefinitionForSave,
    LineDefinitionEntryResourceDefinition
} from './line-definition-entry-resource-definition';
import {
    LineDefinitionEntryNotedAgentDefinition,
    LineDefinitionEntryNotedAgentDefinitionForSave
} from './line-definition-entry-noted-agent-definition';
import {
    LineDefinitionEntryNotedResourceDefinition,
    LineDefinitionEntryNotedResourceDefinitionForSave
} from './line-definition-entry-noted-resource-definition';

export interface LineDefinitionEntryForSave<
    TAgentDef = LineDefinitionEntryAgentDefinitionForSave,
    TResourceDef = LineDefinitionEntryResourceDefinitionForSave,
    TNotedAgentDef = LineDefinitionEntryNotedAgentDefinitionForSave,
    TNotedResourceDef = LineDefinitionEntryNotedResourceDefinitionForSave> extends EntityForSave {
    Direction?: 1 | -1;
    ParentAccountTypeId?: number;
    EntryTypeId?: number;

    AgentDefinitions?: TAgentDef[];
    ResourceDefinitions?: TResourceDef[];
    NotedAgentDefinitions?: TNotedAgentDef[];
    NotedResourceDefinitions?: TNotedResourceDef[];
}

export interface LineDefinitionEntry extends LineDefinitionEntryForSave<
    LineDefinitionEntryAgentDefinition,
    LineDefinitionEntryResourceDefinition,
    LineDefinitionEntryNotedAgentDefinition,
    LineDefinitionEntryNotedResourceDefinition> {
    LineDefinitionId?: number;
    SavedById?: number;
}
